feat: 회원 상세 정보 저장 및 조회 구현

devholic22 opened the pull request:
## 📄 Summary
- [x] 회원가입 시 필요한 회원 정보 속성 정의
- [x] 회원 정보 저장
- [x] 회원 정보 조회
- [x] 회원 정보 수정 & 삭제

## 🙋🏻 More
### 회원 정보 - 구조
* 회원 정보 (로그인에 필요한 정보를 제외한 모든 정보)의 시작 패키지는 `info`로 두었습니다.
* 회원 정보를 크게 `profile`, `option`, `hobby`, `style`로 나누었습니다. (`image`도 있으나, 파일 업로드 과정이 포함되어 현재 개발에서는 클래스만 정의해두었습니다.)
  * `profile`은 `Body` (생년월일, 키, 성별), `Location` (사는 지역 - `city`는 시/도, `sector`는 구를 나타냅니다.), `Position` (latitude: 위도, longitude: 경도), `Job` (직업)을 가집니다.
  * `option`은 `Smoke` (흡연), `Religion` (종교), `Drink` (음주), `Mbti` (MBTI), `Graduate` (최종학력)을 가집니다.
  * `hobby`는 회원 (`Member`)과 등록된 취미 이름 (`HobbyName` - enum) 간의 다대다 구조를 나타내기 위한 브릿지 테이블 엔티티입니다. `style`도 회원과 스타일 이름 (`StyleName` - enum) 간의 다대다 구조를 나타내기 위한 브릿지 테이블 엔티티입니다.
### 회원 정보 - 저장
* 피그마를 봤을 때 각각 속성을 선택하고 최종 저장을 누르는 것으로 생각해서 처음 저장 시 profile, option, hobby, style을 모두 작성하도록 했습니다.
  * profile, option, hobby, style을 저장하는 행위는 각각의 이벤트 핸들러를 만들어 해결하였습니다. 그렇지 않을 경우 infoService에서 각 하위 서비스를 모두 알아야 할 수 밖에 없기 때문입니다.
### 회원 정보 - 조회
* 조회는 이벤트로 이끌어낼 수 없기에, 컨트롤러를 세분화하여 보여주는 식으로 해결하였습니다. 예시로 `profile`을 조회한다면 `/api/info/profile`, 취미를 조회한다면 `/api/info/hobby` 이런 식으로 진행되도록 했습니다.
### 회원 정보 - 수정
* 수정은 새롭게 API를 만들지 않고 처음 만든 저장 API를 그대로 활용하도록 했습니다. 각각의 속성만 (ex: 종교만 선택) 변경하도록 메서드를 만들면 속성이 10개가 넘어감에 따라 변경 API도 많아질 것이라 생각했습니다. 그래서 option과 profile의 경우에는 내용을 수정하고, hobby나 style 같은 경우에는 다대다 구조였기에 기존 내역을 입력받은 내역으로 완전히 교체하도록 구현했습니다. 다만 이 부분은 논의해야 할 부분인 것 같습니다.
> close #7

devholic22 committed:
feat: 프로필 속성 정의

- 취미 (Hobbies, Hobby) 정의
- 나의 스타일 (Styles, Style) 정의
- 프로필 사진 (Images, Image) 정의

devholic22 committed:
feat: 프로필 속성 정의

- 나이 (Age) 정의
- 키 (Height) 정의
- 직업 (Job) 정의
- 지역 (Location) 정의
- 최종학력 (Graduate) 정의
- 성별 (Gender) 정의
- 흡연 (Smoke) 정의
- 음주 (Drink) 정의
- 종교 (Religion) 정의
- MBTI (Mbti) 정의

devholic22 committed:
refactor: 프로필 속성 구조 수정

- 나이, 키, 성별을 MemberBody로 취합
- 선택 사항을 MemberOption으로 취합
- nullable false 조건 추가

devholic22 committed:
refactor: 프로필 속성 구조 수정

- Member 생성 시 Hobbies, Styles, Images 기본 객체 생성되도록 정의
- MemberBody를 MembrProfile로 이름 수정 (로그인에 쓰이지 않는 프로필 정보)
- Location, Job을 MemberProfile로 이동

devholic22 committed:
fix: 로그인 정보 식별 오류 수정

- TokenProvider의 extract를 이메일로 수정
- LoginValidCheckerInterceptor에서 TokenProvider의 이메일을 가진 회원의 id를 저장하도록 수정

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
refactor: 클래스 경로 및 구조 수정

- 기존 도메인에 있던 클래스 파일의 패키지 변경
- Hobbies, Styles 엔티티화
- MemberOption, MemberProfile 대신 Option, Profile로 수정
- Option과 Profile을 담은 겉구조를 info로 변경
- Profile 안에 있는 나이, 키, 성별을 MemberBody에 담도록 수정
- Null 예외처리 추가
- 범위 예외처리 추가
- 속성값 조회 실패 예외처리 추가

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
feat: Repository 정의

- Styles 리포지터리 정의
- Hobbies 리포지터리 정의
- Option 리포지터리 정의
- Profile 리포지터리 정의

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
feat: DTO 정의

- Profile, Option, Hobbies, Styles가 포함된 요청을 가지는 InfoWriteRequest 정의
- 그 외 필요한 세부 DTO 정의

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
feat: 이벤트 처리 구현

- Profile, Option 작성 이벤트 처리 및 Factory 구현
- Hobby, Style 작성 이벤트 처리 구현

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
feat: InfoController 작성

- AuthMember와 InfoWriteRequest를 이용한 컨트롤러 메서드 작성

Co-authored-by: eom-tae-in <eti0728@daum.net>

devholic22 committed:
feat: InfoController 작성

- AuthMember와 InfoWriteRequest를 이용한 컨트롤러 메서드 작성

devholic22 committed:
Merge branch 'feat/7' of https://github.com/sosow0212/atwoz into feat/7

devholic22 committed:
fix: Hobbies, Styles 저장 오류 수정

- EventHandler들을 컴포넌트 등록하여 저장되도록 수정

devholic22 committed:
refactor: 공백 추가

- Styles 생성자 공백 추가

devholic22 committed:
refactor: Hobbies, Styles 엔티티화 해제

- Hobbies, Styles 엔티티 대신 밸류 타입으로 변경
- 관련 Repository 및 이벤트 핸들러 삭제
- Member에서 등록되도록 수정

devholic22 committed:
fix: Option, Profile 중복 저장 오류 해결

- Option, Profile 저장 시 이미 저장된 객체면 내용을 수정하도록 변경

devholic22 committed:
refactor: MemberService 공통 축약

- 멤버 조회 메서드화하여 중복 제거

devholic22 committed:
refactor: Member id Column 속성 삭제

- Member id 이름 원상복구 진행

devholic22 committed:
refactor: Member에서 Hobbies, Styles 삭제

- Member에서 Hobbies, Styles에 대한 의존 삭제

devholic22 committed:
refactor: 이벤트 시제 변경

- 이벤트 과거시제로 수정

devholic22 committed:
refactor: MemberBody 수정

- MemberBody에서 Body로 이름 변경
- 나이 필드를 birthYear로 변경
- birthYear, height 타입 int로 변경
- int 변경에 따라 Null 검증 삭제

devholic22 committed:
refactor: Profile, Option 이벤트 핸들러 이름 변경

- 메서드에 따라 세부 이벤트가 달라지는 것이므로 이벤트 핸들러 이름에서 Wrote 삭제

devholic22 committed:
refactor: NullValueException 삭제

- NullValueException 예외 삭제

devholic22 committed:
refactor: 사이즈 0일 시 isEmpty 사용

- 리스트 사이즈 검증 시 0일 경우 isEmpty로 교체

devholic22 committed:
refactor: Repository 인터페이스 위치 수정

- Repository 인터페이스 위치를 domain으로 수정

devholic22 committed:
refactor: Member - Hobby 구조 수정

- Member와 Hobby를 M:N 구조로 수정
- Hobby의 이름을 enum으로 관리, 데이터베이스 테이블에서 삭제
- 관련 예외 등록

devholic22 committed:
refactor: Member - Style 구조 수정

- Member와 Style을 M:N 구조로 수정
- Style의 이름을 enum으로 관리, 데이터베이스 테이블에서 삭제
- 관련 예외 등록

devholic22 committed:
fix: Option 수정 기능 오류 수정

- Option 내용 수정 시 반영 안되는 오류 수정

devholic22 committed:
feat: Position 추가

- Position 속성 정의
- 관련 검증 로직 추가

devholic22 committed:
feat: 프로필 (Profile) 조회 작성

- ProfileController를 통해 작성한 Profile 조회 개발
- 응답 DTO 정의
- Repository 메서드를 추가하여 Optional 방지

devholic22 committed:
feat: 옵션 (Option) 조회 작성

- OptionController를 통해 작성한 Option 조회 개발
- 응답 DTO 정의
- Repository 메서드를 추가하여 Optional 방지

devholic22 committed:
refactor: OptionController 위치 수정

- OptionController 위치 수정

devholic22 committed:
feat: 회원 취미 조회 작성

- HobbyController를 통해 작성한 Hobby 조회 개발
- 응답 DTO 정의

devholic22 committed:
feat: 회원 스타일 조회 작성

- StyleController를 통해 작성한 Style 조회 개발
- 응답 DTO 정의

devholic22 committed:
refactor: Body 생성 수정

- Member의 Body에 출생년도가 들어가도록 수정
- 나이 최저 계산 기준 수정
- createMemberBody의 내부 타입 Integer에서 int로 변경

devholic22 committed:
refactor: Profile 조회 시 이름 수정

- year 대신 age로 나오도록 수정

devholic22 committed:
refactor: 중복 선택 예외처리 추가

- Hobby, Style 중복 선택 시 예외처리 추가

devholic22 committed:
test: HobbyName 테스트 작성

- HobbyName 테스트 작성

devholic22 committed:
test: StyleName 테스트 작성

- StyleName 테스트 작성

devholic22 committed:
test: Gender 테스트 작성

- Gender 테스트 작성

devholic22 committed:
test: BodyTest 작성

- 정상 케이스 테스트 작성
- 키에 대한 예외 검증 테스트 작성
- 몸무게에 대한 예외 검증 테스트 작성
- 기존 연도 계산을 인터페이스화하여 테스트 용이하게 설정

devholic22 committed:
test: Position 테스트 작성

- Position 테스트 작성

devholic22 committed:
test: ProfileJpaRepository 테스트 작성

- ProfileJpaRepository 테스트 작성
- Body, Job, Location, Position Equals&HashCode 정의
- Body, Job, Location, Position Fixture 생성

devholic22 committed:
test: OptionJpaRepository 테스트 작성

- OptionJpaRepository 테스트 작성
- Option Equals&HashCode 정의

devholic22 committed:
test: HobbyJpaRepository 테스트 작성

- HobbyJpaRepository 테스트 작성

devholic22 committed:
test: StyleJpaRepository 테스트 작성

- StyleJpaRepository 테스트 작성

devholic22 committed:
test: HobbyService 테스트 작성

- HobbyService 테스트 작성
- HobbyFakeRepository 작성
- HobbyService 메서드 이름 수정

devholic22 committed:
test: StyleService 테스트 작성

- StyleService 테스트 작성
- StyleFakeRepository 작성
- StyleService 메서드 이름 수정

devholic22 committed:
test: OptionService 테스트 작성

- OptionService 테스트 작성
- OptionFakeRepository 작성
- OptionFactory 정적 클래스로 변경

devholic22 committed:
test: ProfileService 테스트 작성

- ProfileService 테스트 작성
- ProfileFakeRepository 작성
- FakeYearManager 작성
- Profile Equals&HashCode 정의

devholic22 committed:
test: OptionServiceTest 수정

- OptionFactory 컴포넌트화
- OptionServiceTest save 검증 수정

devholic22 committed:
test: InfoService 테스트 작성

- InfoService 테스트 작성
- 이벤트 발생 여부로 검증 진행

devholic22 committed:
test: HobbyControllerWebMvc 테스트 작성

- HobbyControllerWebMvc 테스트 작성
- REST Docs 연결 작성

devholic22 committed:
test: HobbyControllerWebMvc 수정

- ObjectMapper 삭제

devholic22 committed:
test: StyleControllerWebMvc 테스트 작성

- StyleControllerWebMvc 테스트 작성
- REST Docs 연결 작성

devholic22 committed:
test: OptionControllerWebMvc 테스트 작성

- OptionControllerWebMvc 테스트 작성
- REST Docs 연결 작성

devholic22 committed:
test: ProfileControllerWebMvc 테스트 작성

- ProfileControllerWebMvc 테스트 작성
- REST Docs 연결 작성

devholic22 committed:
test: InfoControllerWebMvc 테스트 작성

- InfoControllerWebMvc 테스트 작성
- REST Docs 연결 작성

devholic22 committed:
refactor: 예외 핸들러에 추가

- info 내부 예외들을 MemberExceptionHandler에 등록

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/hobby/Hobby.java:
```diff
@@ -0,0 +1,36 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(name = "member_hobbies")
+@Entity
+public class Hobby {
```
지금 enum 타입으로 관리하는 클래스의 이름을 Hobby로 변경하고 해당 클래스의 이름을 MemberHobby로 변경하는 것이 좀 더 클래스를 파악하기 쉬울 것 같습니다.

devholic22 replied:
수정 완료하였습니다! 말씀하신대로 MemberHobby로 해야 Member와 Hobby의 구조인 것을 더 쉽게 알 수 있다고 생각합니다 :)

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyName.java:
```diff
@@ -0,0 +1,89 @@
+        if (validSize > MAX_SIZE) {
+            throw new HobbySizeException();
+        }
+    }
```
size에 대한 검증을 메서드로 분리하신 거 같은데 이 메서드도 분리하는 게 좀 더 좋아보입니다.
1. names가 비었는지 검증하는 메서드
2. names의 중복을 검증하는 메서드
3. names의 크기가 최대 size를 넘는지 검증하는 메서드
이렇게 분리를 한 번더 하면 어떨까요?

devholic22 replied:
네 각각 `validateIsNotEmptyCodes`, `validateIsUniqueCodes`, `validateIsNotOverMaxSizeOfHobby`로 분할하였습니다!
검증 로직이 여러 개일 경우 나누는 습관을 들여왔었는데 이 부분에서 작성하지 않았었네요
다음부터는 습관을 다시 지키도록 하겠습니다

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyName.java:
```diff
@@ -0,0 +1,89 @@
+                .filter(hobbyName -> hobbyName.isSame(name))
+                .findFirst()
+                .orElseThrow(HobbyNotFoundException::new);
+    }
```
정적 팩토리 메서드라 네이밍 컨벤션을 따라 매개변수가 1개라 from이라는 네이밍을 주신 것 같습니다. 
하지만 이 메서드의 이름인 from으로 해당 메서드의 기능을 파악할 수 있을까요?
제가 보기엔 이런 상황에서는 좀 더 주체적인 네이밍을 부여하는게 더 좋아보입니다!
저라면 메서드의 이름을 findBy로 하여 Hobbyname.findBy(String.class); 이런식으로 메서드를 호출하도록 할 것 같습니다:)

devholic22 replied:
`from`은 정적 팩터리 메서드의 컨벤션이면서 동시에 "생성"하는 느낌이 있다면, 말씀하신 대로 `findBy`가 "조회"의 느낌을 주니 그게 더 좋은 듯 합니다.
마냥 네이밍 컨벤션을 무조건적으로 따르기보다는 `메서드의 본래 책임을 더 명확히 할 수 있는 이름은 없는지 고민`해보는 습관을 가져보겠습니다.

더 수정하여 `findByCode`로 변경하였습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyRepository.java:
```diff
@@ -0,0 +1,11 @@
+    void save(final Hobby hobby);
+    void saveAll(final List<Hobby> memberHobbies);
+    void deleteHobbiesByMemberId(final Long memberId);
+    List<Hobby> findAllByMemberId(Long memberId);
```
final 키워드 추가해주세요

devholic22 replied:
수정하였습니다!

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/image/Images.java:
```diff
@@ -0,0 +1,19 @@
+@Embeddable
+public class Images {
+
+    @ElementCollection
```
값타입 컬렉션은 값을 변경했을 때 기존의 값을 모두 delete하고 다시 insert하는 단점이 있습니다. 다른 방법을 고민해보는 것이 좋을 것 같습니다.

sosow0212 commented on src/docs/asciidoc/option.adoc:
```diff
@@ -0,0 +1,15 @@
+
+=== Option 조회
+
+==== 요청
```
요청에는 문서가 없네요 어떤 url로 어떤 요청을 하는지 모를 것 같은데
필드가 없다면 http-request.adoc[] 을 사용하는게 좋아보이네요 다른 부분에도 마찬가지지만 요청, 응답 두 부분에 http는 모두 붙여주세요

devholic22 replied:
수정 완료하였습니다
InfoController만 남겨두어서 `info.adoc`에 적용했습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/InfoService.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Transactional
+    public void writeProfile(final Long memberId, final InfoWriteRequest request) {
+        ProfileWriteRequest profileWriteRequest = request.profile();
+        OptionWriteRequest optionWriteRequest = request.option();
```
사소한거긴한데 이건 큰 로직이 없는 부분인데, 변수로 추출하지 말고 이벤트에 보낼 때 매개변수로 request.profile로 태우는 건 어떨까요?

변수 추출한다면 밑에 stream을 사용하는 부분에서 오히려 변수로 추출하는 게 더 나아보여서요

devholic22 replied:
동의합니다! `request.profile()`, `request.option()`으로만 주면 코드가 더 줄어들 수 있고 현재 별다른 작업을 수행하지 않으니 말씀하신 게 맞다고 생각됩니다. 수정 완료하였습니다

devholic22 replied:
생각해보니 말씀하신대로 굳이 변수로 추출할 필요가 없었던 것 같습니다. 이벤트에 바로 넘기는 방식으로 수정하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/application/info/InfoService.java:
```diff
@@ -0,0 +1,42 @@
+    public void writeProfile(final Long memberId, final InfoWriteRequest request) {
+        ProfileWriteRequest profileWriteRequest = request.profile();
+        OptionWriteRequest optionWriteRequest = request.option();
+        List<String> hobbyNames = request.hobbies()
+                .stream()
+                .map(HobbyWriteRequest::hobby)
+                .toList();
+        List<String> styleNames = request.styles()
+                .stream()
+                .map(StyleWriteRequest::style)
+                .toList();
```
이 부분도 각각 메서드 분리 한 번씩 해주면 좋을 것 같아요

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/HobbyWriteRequest.java:
```diff
@@ -0,0 +1,10 @@
+import jakarta.validation.constraints.NotBlank;
+
+public record HobbyWriteRequest(
+
```
매개변수 안에 있어서 한 칸 띄우지 않아도 됩니다!

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/InfoWriteRequest.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.List;
+
+public record InfoWriteRequest(
+
```
1. 공백을 넣은 이유가 뭘까요?
2. validation을 생략하신건가요?

devholic22 replied:
* 위에서 말씀하신 것 처럼 공백을 삭제하였습니다.
* validation을 넣지 않으니 하위 DTO 자체를 넣지 않았을 때 NPE가 발생함을 알았습니다. validation을 추가하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/hobby/HobbyService.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Transactional
+    public void saveMemberHobbies(final Long memberId, final List<String> hobbyNames) {
+        deleteMemberHobbies(memberId);
```
프론트에서 중복으로 해당 메서드를 재사용할 것 같지 않아서 지우는 행위는 업데이트하는 메서드에서 해줘도 될 것 같습니다.

devholic22 replied:
수정 완료하였습니다! 생성과 수정을 동일한 메서드로 이용하도록 했던 판단때문에 이렇게 설계했던 것 같습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/profile/Body.java:
```diff
@@ -0,0 +1,55 @@
+            throw new ProfileRangeException();
+        }
+        return memberAge;
+    }
```
Math.abs()로 currentYear - birthYear를 하신 이유가 뭔가요? (currentYear가 birthYear보다 작을 수가 있나요?)
만약 임의로 birthYear를 선택할 수 있어서 birthYear로 currentYear보다 큰 값을 선택할 수 있다면 이건 따로 예외처리를 해야할 것 같은데 어떻게 생각하시나요?

devholic22 replied:
실제로는 currentYear는 시스템을 이용해서 현재 연도를 가져올 것이고, 프론트에서는 최대 연도를 지정할 것이기에 이러한 경우는 존재하지 않아 `Math.abs`를 없애는 식으로 해도 될 것 같습니다.

이후에는 `유저 입장에서 작성 가능한 요청값들의 범위`에 대해 고민해보고 메서드를 작성하겠습니다.

sosow0212 commented on src/main/java/com/atwoz/member/application/info/hobby/HobbyService.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Transactional
+    public void deleteMemberHobbies(final Long memberId) {
```
saveMemberHobbies에서만 사용하는데 접근제어자가 public으로 되어있습니다.
그리고 지금은 메서드로 나눌 필요가 없어보이네요

devholic22 replied:
단순 메서드 호출은 메서드로 나눌 필요가 없는 게 가독성 측면에서 더 낫다는 것을 다시 깨달았습니다..
접근제어자를 public으로 했던 이유는 `@Transactional` 때문이었는데, 이 또한 메서드를 나누지 않았다면 고민하지 않았던 부분인 것 같습니다. 수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/option/OptionFactory.java:
```diff
@@ -0,0 +1,25 @@
+public class OptionFactory {
+
+    public Option fromRequest(final Long memberId, final OptionWriteRequest request) {
+
```
불필요한 공백있습니다

devholic22 replied:
수정했습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/option/OptionFactory.java:
```diff
@@ -0,0 +1,25 @@
+import com.atwoz.member.domain.info.option.Smoke;
+import org.springframework.stereotype.Component;
+
+@Component
```
1. 해당 클래스는 무엇을 하는 클래스인가요?
2. Component를 통해 빈으로 등록하신 이유가 무엇인가요

devholic22 replied:
* ProfileFactory처럼 Option을 생성하는 데 책임이 있습니다.
* 같은 이유로 빈 등록을 해제하였습니다. 단순 생성인만큼 정적 메서드로 사용해도 충분했는데 사용을 잘못했었습니다.

sosow0212 commented on src/main/java/com/atwoz/member/application/info/option/OptionService.java:
```diff
@@ -0,0 +1,44 @@
+    private final OptionFactory optionFactory;
+
+    @Transactional
+    public void writeOption(final Long memberId, final OptionWriteRequest request) {
```
write인데 업데이트까지 같이 해주는 것 같네요 나누는 것이 낫지 않을까요?

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/option/OptionService.java:
```diff
@@ -0,0 +1,44 @@
+        );
+    }
+
+    private boolean isMemberOptionExist(final Long memberId) {
```
이것도 단일 repository 메서드인데 메서드로 분리할 필요가 없어보입니다

devholic22 replied:
수정 완료했습니다! 말씀하신대로 메서드 호출인데 굳이 분리할 필요가 없었네요..

sosow0212 commented on src/main/java/com/atwoz/member/application/info/profile/ProfileFactory.java:
```diff
@@ -0,0 +1,50 @@
+        return new Body(currentYear, birthYear, height, gender);
+    }
+
+    private Location createLocation(final LocationWriteRequest request) {
+        return new Location(request.city(), request.sector());
+    }
+
+    private Position createPosition(final PositionWriteRequest request) {
+        return new Position(request.latitude(), request.longitude());
+    }
+
+    private Job createJob(final String job) {
+        return new Job(job);
+    }
```
이 부분을 메서드로 나눈 이유가 무엇인가요? `return new Body(~)` 이 부분은 메서드로 나뉘어 있지 않네요.
일관성이 없고, 불필요하게 나눠서 너무 많은 메서드로 가독성이 오히려 떨어지는 것 같습니다

devholic22 replied:
* 처음에는 전달되는 인자 타입이 ProfileWriteRequest의 내부 타입들이기에 이들을 분리하는 것이 좋겠다는 생각을 했었습니다. (`profileWriteRequest.position().latitude()` 이런 식으로 작성하지 않고자 함이었습니다.)
* 하지만 말씀하신대로 메서드가 너무 많이 분리되는 것 같아 하나의 메서드에서 처리하도록 수정하였습니다. (추가로 Request에 대한 로직이 비슷하여, 이를 인터페이스화 하여 통일하였습니다.)

sosow0212 commented on src/main/java/com/atwoz/member/application/info/profile/ProfileService.java:
```diff
@@ -0,0 +1,45 @@
+public class ProfileService {
+
+    private final ProfileRepository profileRepository;
+    private final ProfileFactory profileFactory;
```
Factory가 필요한지 의문이 드네요.
생성의 책임을 나누는 클래스인 것 같은데 그 역할에 로직이 있는 것도 아니고 빈으로 등록해서 주입 받아 사용하는 것은 크게 의미가 없어보입니다.
 그냥 해당 클래스에서 만드는 메서드를 만들거나 해도 되지 않을까요?

devholic22 replied:
* 빈으로 등록할 필요가 없는 것에 동의합니다. (원래 의도는 정적 메서드로 만들고자 했는데, 컴포넌트로 등록해버린 것 같습니다.)
* Factory 자체의 필요성에 대해서는 Profile을 생성하는 단계가 너무 많다 보니 (Body를 만들고, Location을 만드는 등) 바로 요청 dto를 엔티티로 변환해주는 별도의 클래스가 있는 게 좋겠다는 생각이 들어 만든 것이었는데, (ProfileService에서 dto를 엔티티로 만든다면 내부적으로 많은 메서드가 필요할 것으로 예상했었습니다.) 혹시 재윤님은 Factory를 보통 어떨 때 만드시는지 여쭤봐도 될까요?

sosow0212 commented on src/main/java/com/atwoz/member/config/AuthConfig.java:
```diff
@@ -33,7 +32,7 @@ public void addInterceptors(InterceptorRegistry registry) {
         return new PathMatcherInterceptor(parseMemberIdFromTokenInterceptor)
                 .excludePathPattern("/**", OPTIONS)
-                .addPathPatterns("/admin/**", ANY);
+                .addPathPatterns("/**", ANY);
```
`/**`로 설정하면 모든 url에 대해 인터셉터를 할텐데 이렇게 하신 이유가 무엇인가요?

devholic22 replied:
일단 이렇게 하지 않으면 토큰을 주더라도 `로그인 정보를 찾을 수 없습니다` 예외가 발생해서 변경했었습니다.
하지만 말씀하신 내용을 보니, `인터셉터란 무엇인가?`에 대한 이해를 하지 않은 채 문제를 해결하려고 했던 것 같습니다.

인터셉터란 컨트롤러의 핸들러를 호출하기 전후 추가적인 작업을 원할 때 사용되는 것으로써, 현재 `ParseMemberIdFromTokenInterceptor`는 `LoginValidCheckerInterceptor`와 연계하여 사용자 정보를 인증하기 위해 사용됩니다.

따라서 전체 url에 대한 인터셉터에서, 현재 작업과 관련된 `/api/info/**` 로 변경 완료하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/domain/auth/TokenProvider.java:
```diff
@@ -4,5 +4,5 @@ public interface TokenProvider {
     String createTokenWith(final String email);
 
-    Long extract(final String token);
+    String extract(final String token);
```
* re
이거 rebase하면서 바뀐 것인가요? 아니면 직접 바꾸신건가요?

devholic22 replied:
직접 바꿨었습니다. `createTokenWith` 메서드에서 claims로 id가 아니라 email만 담고 있기 때문에, Long id를 이끌어낼 수 없다고 판단했습니다.

sosow0212 commented on src/main/java/com/atwoz/member/ui/auth/interceptor/LoginValidCheckerInterceptor.java:
```diff
@@ -24,8 +27,10 @@ public boolean preHandle(final HttpServletRequest request,
-        Long memberId = tokenProvider.extract(token);
-        authenticationContext.setAuthentication(memberId);
+        String memberEmail = tokenProvider.extract(token);
+        Member findMember = memberRepository.findByEmail(memberEmail)
+                .orElseThrow(RuntimeException::new);
```
이 부분도 필요하나요? 조회를 최소화 하려고 id를 준 것인데 무조건 조회가 발생하는 것 같네요
토큰에서 email을 찾고 그걸로 member를 조회하고 id를 심는게 아무런 의미도 없어보여요
토큰에서 id값이 담겨있다면 그걸로 이미 가입된 멤버임을 알 수 있지 않나요?

devholic22 replied:
토큰에 id가 담기려면 `createTokenWith`에서 id가 인자로 전달되어야 하는데, email로 전달되어 있습니다.
만약 `createTokenWith` 메서드에서 id 값이 제공된다면 원래 작성하신대로 해도 될 것 같습니다.
다만 지금 기능을 개발할 때는 소셜 로그인이 email만 제공할 수 있다고 이해했어서 구조를 바꾸었었습니다.

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyName.java:
```diff
@@ -0,0 +1,89 @@
+    GOLF("골프"),
+    THEATER("공연 · 전시회관람"),
+    WRITE("글쓰기"),
+    GAME("PC · 모바일 게임"),
```
`·` 기호를 저렇게 valaue로 하신 이유가 있을까요? 프론트에서 따로 코드로 전달된다고 그랬는데 아무런 의미가 없어보입니다 ㅠ

devholic22 replied:
프론트에서 전달할 코드를 기준으로 조회되도록 수정하였고, Style, Job도 동일한 기준으로 조회되도록 수정하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyName.java:
```diff
@@ -0,0 +1,89 @@
+        if (validSize > MAX_SIZE) {
+            throw new HobbySizeException();
+        }
+    }
```
1.  외부에서 사용하지 않습니다 접근제어자 수정해주세요
2. validateEmptySizeOfHobbies
3. validateDuplicatedHobbies
4. validateOverMaxSizeOfHobby

devholic22 replied:
1. 수정했습니다. 메서드명을 `validateCodes`로 변경했습니다.
2. `validateIsNotEmptyCodes`로 변경하였습니다.
3. `validateIsUniqueCodes`로 변경하였습니다.
4. `validateIsNotOverMaxSizeOfHobby`로 변경하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/hobby/HobbyRepository.java:
```diff
@@ -0,0 +1,11 @@
+    void save(final Hobby hobby);
+    void saveAll(final List<Hobby> memberHobbies);
+    void deleteHobbiesByMemberId(final Long memberId);
+    List<Hobby> findAllByMemberId(Long memberId);
```
매개변수에 여기는 final이 없네요 모두 없어도 되는데 통일해주세요

devholic22 replied:
수정 완료했습니다!

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/option/Option.java:
```diff
@@ -0,0 +1,64 @@
+    @Enumerated(value = EnumType.STRING)
+    private Graduate graduate;
+
+    public Option(final Long memberId, final Smoke smoke, final Religion religion,
```
매개변수를 3개씩 나눈 이유가 무엇인가요?? 개행을 하나씩 넣어주는게 나아보여요 저런 컨벤션은 없었습니다 ㅠ

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/profile/Body.java:
```diff
@@ -0,0 +1,55 @@
+
+    private void validateHeight(final int height) {
+        if (height < MIN_HEIGHT || MAX_HEIGHT < height) {
+            throw new ProfileRangeException();
```
예외 메시지가 너무 단순합니다. 구체적으로 다른 Exception을 사용하는게 좋을 것 같습니다
그리고 테스트 코드에서도 합쳐서 사용중인데 분리해서 작성해주세요

devholic22 replied:
ProfileRangeException을 삭제하고 AgeRangeException, HeightRangeException, LatitudeRangeException, LongitudeRangeException으로 세분화하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/profile/Position.java:
```diff
@@ -0,0 +1,43 @@
+    private static final int MIN_LONGITUDE = -180;
+    private static final int MAX_LONGITUDE = 180;
+
+    private double latitude;
```
위도 경도면 BigDecimal로 바꾸는 건 어떨까요? double과 무슨 차이가 있을까요?

devholic22 replied:
`double`, `float`보다 `BigDecimal`이 더 미세한 숫자의 변동을 허용하지 않아 금융 등 민감한 데이터에서 다룸을 알았습니다.
위도 경도 또한 특정 거리만큼 집합을 만들거나 할 수 있기에 말씀하신대로 `BigDecimal`을 사용하는 것이 좋을 듯 합니다.
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/domain/info/profile/year/YearManager.java:
```diff
@@ -0,0 +1,6 @@
+package com.atwoz.member.domain.info.profile.year;
+
+public interface YearManager {
```
현재 플로우는
1. factory에서 사용중이고
2. 외부 서비스에서 getCurrentYear로 생성하고
3. 도메인을 생성할 때 이 값을 넘겨주며 생성하고
4. 생성시 나이를 계산

이런 구조로 되어있는데, 도메인 서비스로 도메인 생성자나 다른 부분에 넘겨서 한 번에 해주도록 하는게 나아보입니다
외부에 노출될 필요가 없을 것 같아서요

devholic22 replied:
YearManager의 사용을 도메인에서 하도록 수정하였습니다. (생성자에 인터페이스 전달)

sosow0212 commented on src/main/java/com/atwoz/member/infrastructure/info/option/OptionRepositoryImpl.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public boolean isExistMemberOption(final Long memberId) {
+        return findByMemberId(memberId).isPresent();
+    }
```
이렇게 만든 이유가 있을까요?
jdbc를 사용하든, jpa에서 다른 메서드를 사용해서 더 효율적으로 쿼리를 날릴 수 있습니다 ex. count query...

devholic22 replied:
수정 완료하였습니다. 성능에 대해 깊게 고민하지 않았던 것 같습니다.
처음에는 `count(*)`을 이용하여 수정하였으나, 이후 `exists`가 더 빠르다는 것을 알게 되어 재수정하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/HobbyController.java:
```diff
@@ -0,0 +1,32 @@
+@RequiredArgsConstructor
+@RequestMapping("/api/info/hobby")
+@RestController
+public class HobbyController {
```
info부분에서 다 해결될 것 같은데 따로 hobby, option 등등을 컨트롤러랑 서비스를 나눈 이유를 잘 모르겠습니다. 
그리고 Restful 계층 구조도 안지켜지고 있는 것 같습니다.

devholic22 replied:
* Querydsl을 활용함으로써 InfoController를 제외한 나머지 컨트롤러들을 모두 삭제 완료하였습니다.
* 다만 서비스는 만약 서비스를 나누지 않을 경우 InfoService에서 모든 리포지터리 (Profile, Option, MemberHobby, MemberStyle)에 대해 의존하고 작업을 수행해야 하는데, 그러기보다는 이벤트로 복잡성을 분리하는 게 더 낫겠다는 생각을 했었습니다.

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/InfoController.java:
```diff
@@ -0,0 +1,29 @@
+        return ResponseEntity.ok()
+                .build();
+    }
+
```
불필요한 공백이 있습니다

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/InfoController.java:
```diff
@@ -0,0 +1,29 @@
+    public ResponseEntity<Void> writeInfo(@AuthMember final Long memberId,
+                                          @Valid @RequestBody final InfoWriteRequest request) {
+        infoService.writeProfile(memberId, request);
+        return ResponseEntity.ok()
```
생성은 201 응답 주기로 한 것 같습니다!

devholic22 replied:
수정 완료하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/StyleController.java:
```diff
@@ -0,0 +1,32 @@
+    public ResponseEntity<List<StyleSearchResponse>> findMemberStyles(@AuthMember Long memberId) {
+        List<StyleSearchResponse> styles = styleService.findMemberStyles(memberId)
+                .stream()
+                .map(Style::getStyleName)
+                .map(styleName -> new StyleSearchResponse(styleName.getName()))
```
이것도 한 번에 변경할 수 있을 것 같네요

devholic22 replied:
아래처럼 DTO에서 변환하도록 수정 완료했습니다!

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/StyleController.java:
```diff
@@ -0,0 +1,32 @@
+    private final StyleService styleService;
+
+    @GetMapping
+    public ResponseEntity<List<StyleSearchResponse>> findMemberStyles(@AuthMember Long memberId) {
```
여기 매개변수에는 final이 생략되어 있습니다

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/HobbyServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@ExtendWith(MockitoExtension.class)
+public class HobbyServiceTest {
```
테스트 클래스는 public 접근 제어자 붙일 필요가 없습니다.

devholic22 replied:
수정하였습니다! 다른 클래스 코드들에서도 모두 제거하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/application/info/HobbyServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@ExtendWith(MockitoExtension.class)
```
Mockito를 사용하는 이유가 무엇인가요? 이미 Fake로 테스트하는 것 같습니다

devholic22 replied:
Mickito의 사용 이유는 when, verify 등 행위 검증을 할 때 정의해두기 위함인데, 말씀하신 대로 Fake를 이미 이용하기 때문에 없애도 될 것 같습니다. 수정 완료하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+    private InfoService infoService;
+
+    @Autowired
+    ApplicationEvents events;
```
접근 제어자가 없습니다

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@SpringBootTest
+@RecordApplicationEvents
+@ExtendWith(MockitoExtension.class)
```
해당 어노테이션들을 각각 붙인 이유가 궁금합니다!

devholic22 replied:
* MockitoExtension: 앞서 말씀드린대로 잘못 작성했었던 것 같습니다.. (mocking을 할 필요가 없는데 붙였었습니다.)
* RecordApplicationEvents: 이벤트 발생을 시키기 때문에 이벤트 발생 횟수를 검증하기 위해 적용하였습니다.
* SpringBootTest: 서비스 테스트이긴 하나, ApplicationEvents를 할당받기 위해서는 SpringBootTest를 붙여야 NullPointerException이 발생하지 않는다는 것을 알게 되어 적용했습니다. (이후 IntegrationHelper를 적용함으로써 제거하였습니다.)

sosow0212 commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+        // given
+        Long memberId = 1L;
+
+        int birthYear = 2000;
+        int height = 171;
+        String gender = "남성";
+        LocationWriteRequest location = new LocationWriteRequest("서울시", "강남구");
+        double latitude = 70.5;
+        double longitude = 140.3;
+        PositionWriteRequest position = new PositionWriteRequest(latitude, longitude);
+        String job = "개발자";
+        ProfileWriteRequest profileWriteRequest = new ProfileWriteRequest(
+                birthYear,
+                height,
+                gender,
+                location,
+                position,
+                job
+        );
```
지금처럼 정보가 많다면 Fixture를 사용하는게 다른 계층에서 사용할 때에도 테스트 로직에 더욱 가독성있게 볼 수 있어서 좋을 것 같습니다.

devholic22 replied:
Fixture 작성 완료하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+        int hobbyWroteEventCount = (int) events.stream(HobbyWroteEvent.class).count();
+        int styleWroteEventCount = (int) events.stream(StyleWroteEvent.class).count();
+
+        assertSoftly(softly -> {
```
정보가 저장되었는지 확인은 안해도 되나요?

devholic22 replied:
`모든_정보를_조회한다()`를 통해 검증하는 방식으로 수정하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/application/info/OptionServiceTest.java:
```diff
@@ -0,0 +1,92 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@ExtendWith(MockitoExtension.class)
```
여기도 mockito를 사용한 이유가 무엇인가요?

devholic22 replied:
서비스 테스트들에서 습관적으로 MockitoExtension을 붙인 것 같습니다. 삭제 완료하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/application/info/ProfileServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+
+        // then
+        assertSoftly(softly -> {
+            softly.assertThat(profileRepository.isExistMemberProfile(memberId)).isTrue();
+            softly.assertThat(profileRepository.findByMemberId(memberId)).isPresent();
```
이 두 개는 겹치는 거 같네요

devholic22 replied:
두 번째인 `softly.assertThat(profileRepository.findByMemberId(memberId)).isPresent();`만 남겨두는 식으로 수정 완료하였습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/infrastructure/info/hobby/HobbyJpaRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+    void deleteByMemberId(final Long memberId);
+
+    List<Hobby> findAllByMemberId(Long memberId);
```
final 키워드 추가해주세요

devholic22 replied:
수정하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/ProfileServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+        assertSoftly(softly -> {
+            softly.assertThat(profileRepository.isExistMemberProfile(memberId)).isTrue();
+            softly.assertThat(profileRepository.findByMemberId(memberId)).isPresent();
+            softly.assertThat(profileRepository.findByMemberId(memberId).get()).isEqualTo(expectedProfile);
```
equals를 id로 재정의 했다면 id 값만 같으면 true를 반환할텐데 값도 같이 검증하는게 중요할 것 같습니다. 다른 메서드를 통해서 검증해보면 좋을 것 같습니다

devholic22 replied:
`usingRecursiveComparison`, `ignoringFields("id")`을 사용하여 수정 완료하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/fixture/info/BodyFixture.java:
```diff
@@ -0,0 +1,15 @@
+
+public class BodyFixture {
+
+    private static final int CURRENT_YEAR = 2024;
+    private static final int BIRTH_YEAR = 2000;
+    private static final int HEIGHT = 171;
```
fixture에서 다른 클래스에선 변수로 바로 할당해주거나 하는데 이 부분에서만 상수화한 이유가 있을까요?

devholic22 replied:
다른 클래스들과의 통일성을 생각하지 못했었습니다. 다른 메서드들처럼 변수로 할당해주는 게 좋을 것 같아 수정하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/infrastructure/info/HobbyFakeRepository.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public void save(final Hobby hobby) {
+        Hobby newHobby = new Hobby(hobby.getMemberId(), hobby.getHobbyName());
+        map.put(id, newHobby);
```
이렇게 되면 저장된 newHobby에서는 id 값이 null일 것 같아요

devholic22 replied:
수정하였습니다. 추가로 Style 또한 null이 저장되지 않도록 변경했습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/ui/info/OptionController.java:
```diff
@@ -0,0 +1,28 @@
+    private final OptionService optionService;
+
+    @GetMapping
+    public ResponseEntity<OptionSearchResponse> findOption(@AuthMember Long memberId) {
```
final 붙여주세요

devholic22 replied:
수정하였습니다! 추가로 다른 컨트롤러들에서도 `final`이 붙여져있지 않음을 알게 되었습니다.. 이러한 부분이 말씀하신 `조금만 신경써도 수정할 수 있는 내용`인 것 같습니다. 기본적인 것을 많이 못 지켰는데, 이야기 해 주셨듯 다음 작업들에서는 제출 전 파일들을 점검해보는 시간을 더 길게 가지겠습니다.

sosow0212 commented on src/test/java/com/atwoz/member/ui/info/HobbyControllerWebMvcTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private HobbyService hobbyService;
+
+    @MockBean
+    private AuthArgumentResolver authArgumentResolver;
+
+    @MockBean
+    private ParseMemberIdFromTokenInterceptor parseMemberIdFromTokenInterceptor;
+
+    @MockBean
+    private LoginValidCheckerInterceptor loginValidCheckerInterceptor;
+
+    @MockBean
+    private OAuthArgumentResolver oAuthArgumentResolver;
```
test / helper 패키지에 가면 MockBeanInjection 클래스가 있습니다 여기에 MockBean을 모두 넣고 상속받아 사용하면 캐싱돼서 빨라집니다 `@MockBean(JpaMetamodelMappingContext.class)`도 해당 클래스에 같이 넣어주세요

devholic22 replied:
수정 완료하였습니다! 다른 클래스들도 동일하게 변경하였습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/ui/info/StyleController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @GetMapping
+    public ResponseEntity<List<StyleSearchResponse>> findMemberStyles(@AuthMember Long memberId) {
+        List<StyleSearchResponse> styles = styleService.findMemberStyles(memberId)
+                .stream()
+                .map(Style::getStyleName)
+                .map(styleName -> new StyleSearchResponse(styleName.getName()))
+                .toList();
```
컨트롤러에서 직접 변환해주지 말고 dto에서 해주는게 좋아보입니다

devholic22 replied:
`StyleSearchResponse(final Style style)`을 통해 DTO에서 변환하도록 수정했습니다! 더불어 `HobbySearchResponse`도 같은 맥락으로 수정 완료하였습니다.

devholic22 replied:
`StyleSearchResponse.from` 메서드를 만듦으로써 수정하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/ui/info/HobbyControllerWebMvcTest.java:
```diff
@@ -0,0 +1,86 @@
+        // given
+        Long memberId = 1L;
+
+        when(oAuthArgumentResolver.supportsParameter(any())).thenReturn(true);
+        when(parseMemberIdFromTokenInterceptor.preHandle(any(), any(), any())).thenReturn(true);
+        when(loginValidCheckerInterceptor.preHandle(any(), any(), any())).thenReturn(true);
+        when(authArgumentResolver.supportsParameter(any())).thenReturn(true);
+        when(authArgumentResolver.resolveArgument(any(), any(), any(), any())).thenReturn(memberId);
```
요청 보낼 때 헤더값에 토큰을 넣으면 이런 부분 모두 필요 없습니다 어차피 문서화하려면 헤더에 토큰을 넣는 것까지 테스트 해야하는데 바꿔야할 것 같습니다

devholic22 replied:
완료했습니다! 다른 컨트롤러들에도 동일한 방식으로 수정하였습니다.

sosow0212 reviewed (changes requested):
File Changed가 100개가 넘어서 리뷰하는데 너무 힘들었습니다 ㅠ
다음부터는 작업이 길어질 것 같으면 pr을 나눠서 제출하는게 더 좋을 것 같습니다!
불필요하게 나눈 부분도 있는 것 같고 설계하신 부분 중에 몇 가지가 궁금해서 코멘트 남겼습니다!

eom-tae-in commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@SpringBootTest
```
통합 테스트를 할 때 RANDOM PORT 설정을 하는 이유가 뭘까요? 찾아보시고,
이 테스트는 서비스 테스트인데 왜 @SpringBootTest를 하는 걸까요?

devholic22 replied:
* 통합 테스트 시 RANDOM PORT를 설정하는 이유는 테스트 간 격리를 위해서입니다. (포트 충돌 방지) 또한 애플리케이션을 실행하고 있을 시점에 테스트를 돌릴 때에도 기본 포트 (8080)와 충돌되지 않기 위해 사용하기도 합니다.
* InfoServiceTest에서는 이벤트 발생을 검증하기 위해 `RecordApplicationEvents`를 사용해야 하는데, 이것은 스프링 컨텍스트에서 제공받아야 합니다. 따라서 `@SpringBootTest`가 없을 때 이벤트 검증을 시도한다면 ApplicationEvents의 값에 대해 NullPointerException이 발생해서 `@SpringBootTest`를 붙였었습니다.

eom-tae-in commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+    @Autowired
+    ApplicationEvents events;
+
+    @BeforeEach
+    void init() {
+        infoService = new InfoService();
+    }
```
이걸 여기다가 왜 붙였나요?

devholic22 replied:
`@Autowired`가 있었음에도 `@BeforeEach`를 이용하여 또 할당하였었습니다. 수정하였습니다.

eom-tae-in commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+                hobbyWriteRequests,
+                styleWriteRequests
+        );
+
```
given 절의 설정이 너무 긴 것 같습니다. 불필요한 공백도 제거해주세요~

devholic22 replied:
* 공백 제거 완료하였습니다.
* Fixture를 둠으로써 해결하였습니다.

eom-tae-in commented on src/main/java/com/atwoz/member/domain/info/profile/year/YearManagerImpl.java:
```diff
@@ -0,0 +1,14 @@
+package com.atwoz.member.domain.info.profile.year;
+
```
계층 구조상 현재 추상화한 인터페이스 클래스와 같은 도메인 위치에 있는데 이게 과연 올바른 계층 구조일까요?

devholic22 replied:
말씀하신 게 맞습니다.. YearManager 구현체는 infrastructure, 인터페이스만 domain에 관리되도록 수정하였습니다.

eom-tae-in reviewed (changes requested):
리뷰했습니다. 내용 반영해서 수정해주세요.
+ 왜 인수테스트는 작성하지 않으신건가요?? 
+ 조금만 신경써도 수정할 수 있는 내용이 많았습니다. 다음부터는 이런 부분 신경써서 해주세요

devholic22 committed:
docs: API 문서들에 HTTP request 추가

- API 문서들에 HTTP request 추가

devholic22 committed:
refactor: DTO 인자 공백 삭제

- DTO 인자 선언 시의 첫 공백 삭제

devholic22 committed:
refactor: HobbyName 검증 메서드 분할

- 검증 방식에 따른 세부 메서드 분할

devholic22 committed:
refactor: HobbyRepository final 인자 수정

- 모든 메서드 인자에 final 속성이 붙도록 수정

devholic22 committed:
refactor: OptionFactory 공백 삭제

- OptionFactory 불필요한 공백 삭제

devholic22 committed:
refactor: InfoController 공백 삭제

- InfoController 공백 삭제

devholic22 committed:
refactor: HobbyJpaRepository final 인자 수정

- 모든 메서드 인자에 final 속성이 붙도록 수정

devholic22 committed:
refactor: 컨트롤러 memberId final 인자 수정

- 컨트롤러들에 있는 memberId들에 final 속성이 붙도록 수정

devholic22 committed:
refactor: Body 나이 계산 수정

- 계산 시 절댓값 연산 삭제

devholic22 committed:
refactor: Option, Profile 메서드 인자 줄내림

- Option, Profile 메서드 인자 줄내림 적용

devholic22 committed:
refactor: 테스트 파일 public 제어자 삭제

- 테스트 파일의 public 제어자 삭제

devholic22 committed:
refactor: 서비스 테스트의 MockitoExtension 제거

- 서비스 테스트에서의 MockitoExtension 어노테이션 제거

devholic22 committed:
refactor: InfoServiceTest 필드 제어자 추가

- InfoServiceTest의 ApplicationEvents 접근 제어자 추가

devholic22 committed:
refactor: ProfileServiceTest 중복 검증 삭제

- isTrue 대신 isPresent로만 검증하도록 수정

devholic22 committed:
refactor: YearManager 패키지 위치 수정

- YearManager 구현체는 infrastructure에 관리되도록 수정
- YearManager 인터페이스만 domain에 관리되도록 수정

devholic22 committed:
refactor: HobbyName 조회 메서드명 수정

- 메서드의 기능을 더 잘 파악하기 위해 from 대신 findBy로 변경

devholic22 committed:
refactor: StyleName 조회 메서드명 수정

- 메서드의 기능을 더 잘 파악하기 위해 from 대신 findBy로 변경

devholic22 committed:
refactor: Position BigDecimal 교체

- Position의 double 속성을 BigDecimal로 교체
- 관련 테스트 및 코드 수정

devholic22 committed:
refactor: ParseMemberIdFromTokenInterceptor url 수정

- ParseMemberIdFromTokenInterceptor 적용 url 수정

devholic22 committed:
refactor: Option 존재 여부 메서드 성능 개선

- OptionRepository의 isExistMemberOption을 쿼리로 개선

devholic22 committed:
refactor: Profile 존재 여부 메서드 성능 개선

- ProfileRepository의 isExistMemberOption을 쿼리로 개선

devholic22 committed:
refactor: HobbyFakeRepository id 저장 수정

- HobbyFakeRepository 저장 시 id도 저장되도록 빌더 패턴 이용

devholic22 committed:
refactor: StyleFakeRepository id 저장 수정

- StyleFakeRepository 저장 시 id도 저장되도록 빌더 패턴 이용

devholic22 committed:
refactor: DTO 정적 팩터리 메서드 추가

- HobbySearchResponse에서 엔티티를 변환하도록 수정

devholic22 committed:
refactor: DTO 정적 팩터리 메서드 추가

- StyleSearchResponse에서 엔티티를 변환하도록 수정

devholic22 committed:
refactor: MockBean 캐싱 처리 적용

- 각 컨트롤러들의 MockBean을 MockBeanInjection에 두어 캐싱되도록 수정

devholic22 committed:
refactor: InfoService 메서드 분리

- 변수를 추출하지 않고 직접 전달하도록 수정
- hobbyNames, styleNames 메서드 분리

devholic22 committed:
refactor: MockBean 어노테이션 삭제

- MockBeanInjection에서 이미 MockBean 어노테이션을 가지므로 삭제 조치

devholic22 committed:
refactor: StyleControllerWebMvcTest 헤더 정보 추가

- 불필요한 when 삭제
- REST Docs에 토큰 요청 정보 추가

devholic22 committed:
refactor: ProfileControllerWebMvcTest 헤더 정보 추가

- 불필요한 when 삭제
- REST Docs에 토큰 요청 정보 추가

devholic22 committed:
refactor: OptionControllerWebMvcTest 헤더 정보 추가

- 불필요한 when 삭제
- REST Docs에 토큰 요청 정보 추가

devholic22 committed:
refactor: InfoControllerWebMvcTest 헤더 정보 추가

- 불필요한 when 삭제
- REST Docs에 토큰 요청 정보 추가

devholic22 committed:
refactor: HobbyControllerWebMvcTest 헤더 정보 추가

- 불필요한 when 삭제
- REST Docs에 토큰 요청 정보 추가

devholic22 committed:
refactor: 하위 DTO들의 validation 적용

- 하위 DTO들에 validation 적용하여 null 및 기타 검증되도록 수정

devholic22 committed:
refactor: OptionService 메서드 분리 해제

- OptionService의 isMemberOptionExist 메서드 분리 해제

devholic22 committed:
refactor: Hobby 코드 추가

- Hobby의 코드로 관리되도록 수정

devholic22 committed:
refactor: Style 코드 추가

- Style의 코드로 관리되도록 수정

devholic22 committed:
refactor: Job enum으로 변경

- Job을 enum으로 변경
- 직업 코드로 Job을 조회하도록 수정
- 관련 예외 추가
- Job 테스트 작성

devholic22 committed:
refactor: Hobby 응답 코드로 수정

- Hobby 응답이 코드로 표현되도록 수정

devholic22 committed:
refactor: Style 응답 코드로 수정

- Style 응답이 코드로 표현되도록 수정

devholic22 committed:
refactor: 값 범위 예외 분할

- 나이, 키, 위도, 경도 범위 예외를 각각 분할
- 테스트 코드 수정

devholic22 committed:
refactor: info 작성 상태 코드 변경

- info 작성 시의 상태 코드 변경

devholic22 committed:
refactor: InfoServiceTest beforeEach 삭제

- InfoServiceTest beforeEach 삭제

devholic22 committed:
refactor: Factory 정적 클래스화

- ProfileFactory 정적 클래스화 및 메서드명 변경
- YearManager 인터페이스 전달 방식으로 수정
- OptionFactory 정적 클래스화 및 메서드명 변경

devholic22 committed:
refactor: YearManager 생성자로 이동

- YearManager 사용을 도메인에서 하도록 수정

devholic22 committed:
refactor: InfoService 메서드명 변경

- Profile과의 혼동을 막기 위해 메서드명 변경

devholic22 committed:
refactor: Hobby, Style 요청 이름 변경

- Hobby, Style을 코드로 조회하므로 Name을 Code로 변경

devholic22 committed:
refactor: BodyFixture 변수 할당으로 수정

- BodyFixture 메서드 상수 대신 변수를 사용하도록 수정

devholic22 committed:
refactor: 테스트 데이터 Fixture 추가

- 필요한 데이터 정의를 줄이기 위해 요청 DTO Fixture 추가

devholic22 committed:
refactor: 테스트 데이터 Fixture 적용

- ProfileServiceTest, OptionServiceTest Fixture 적용
- 취미, 스타일 given 데이터 실제 값으로 수정

devholic22 committed:
refactor: Info 조회 모델 생성 (Querydsl)

- Querydsl을 사용하여 전체 정보 파악하도록 수정
- HobbyController, StyleController, ProfileController, OptionController 삭제
- InfoService 조회 메서드 작성
- 조회 DTO 작성
- 관련 테스트 작성, 테스트 Fixture 생성

devholic22 committed:
refactor: Profile, Option 저장 값 비교 방식 수정

- Profile, Option 저장 검증 시 usingRecursiveComparison 사용하도록 수정

devholic22 committed:
refactor: Image 삭제

- Image 개발은 다른 PR에서 진행하도록 하기 위해 파일 삭제 조치

devholic22 committed:
refactor: Hobby, HobbyName 이름 변경

- HobbyName을 Hobby로 이름 변경
- Hobby를 MemberHobby로 이름 변경

devholic22 committed:
refactor: InfoRepositoryImpl 용어 변경

- InfoRepositoryImpl QHobby를 QMemberHobby로 변경

devholic22 committed:
refactor: PropertyNotFoundException 삭제

- PropertyNotFoundException 삭제
- 각각의 세부 예외로 변경

devholic22 committed:
refactor: Job 메서드, 예외 이름 수정

- Job 조회 메서드명 findByCode로 수정
- Job 조회 실패 예외 이름 JobInvalidException으로 수정 및 등록

devholic22 committed:
refactor: Style, StyleName 이름 변경

- StyleName을 Style로 이름 변경
- Style를 MemberStyle로 이름 변경

devholic22 committed:
refactor: MemberStyle 인자 이름 변경

- 기존 Style 인자를 MemberStyle로 변경

devholic22 committed:
refactor: Hobby, Style 조회 실패 예외 메서드명 변경

- Hobby 조회 실패 예외 메서드명 변경, 내용 수정
- Style 조회 실패 예외 메서드명 변경, 내용 수정

devholic22 committed:
refactor: InfoRepositoryImpl 용어 변경

- InfoRepositoryImpl QStyle을 QMemberStyle로 변경

devholic22 committed:
test: Drink 테스트 작성

- Drink 테스트 작성

devholic22 committed:
test: Graduate 테스트 작성

- Graduate 테스트 작성

devholic22 committed:
test: Mbti 테스트 작성

- Mbti 테스트 작성

devholic22 committed:
test: Religion 테스트 작성

- Religion 테스트 작성

devholic22 committed:
test: Smoke 테스트 작성

- Smoke 테스트 작성

devholic22 committed:
test: Profile 테스트 작성

- Profile 테스트 작성 (수정 검증)
- 각 Fixture 메서드 생성, SuppressWarnings 추가

devholic22 committed:
test: Option 테스트 작성

- Option 테스트 작성 (수정)
- 전부 enum 타입이기에 Fixture 생성 대신 직접 생성 방식 선택

devholic22 committed:
refactor: MemberStyleJpaRepositoryTest 이름 수정

- Member 중복 작성 삭제

devholic22 committed:
refactor: MemberStyleJpaRepositoryTest 변수 수정

- map 안에서의 변수명 변경

devholic22 committed:
refactor: MemberHobbyJpaRepositoryTest 변수 수정

- map 안에서의 변수명 변경

devholic22 committed:
refactor: MemberHobbyFakeRepository 변수 수정

- forEach 안에서의 변수명 변경

devholic22 committed:
refactor: MemberStyleFakeRepository 변수 수정

- 저장 변수명 변경
- forEach 안에서의 변수명 변경

devholic22 committed:
refactor: ProfileJpaRepositoryTest ignoringFields 추가

- ProfileJpaRepositoryTest 비교 시 ignoringFields 추가

devholic22 committed:
refactor: OptionJpaRepositoryTest ignoringFields 추가

- OptionJpaRepositoryTest 비교 시 ignoringFields 추가

devholic22 committed:
refactor: Hobby 변수 및 메서드 수정

- filter에서의 변수명 수정
- isSameCode로 메서드명 변경, 진행 방식 수정

devholic22 committed:
refactor: Style 변수 및 메서드 수정

- filter에서의 변수명 수정
- isSameCode로 메서드명 변경, 진행 방식 수정

devholic22 committed:
refactor: Enumerated 위치 수정

- 다른 클래스들과의 통일을 지키기 위해 Enumerated 위치 수정

devholic22 committed:
refactor: NoArgsConstructor 위치 수정

- NoArgsConstructor 위치 수정

devholic22 committed:
refactor: exists 쿼리 성능 개선 적용

- exists를 활용하여 (스프링 데이터 JPA 메서드명) 조회되도록 수정

devholic22 committed:
refactor: 이벤트 RequiredArgsConstructor 통일

- 이벤트들의 통일을 위해 RequiredArgsConstructor 추가

devholic22 committed:
refactor: OptionEventHandler 공백 추가

- 다른 이벤트 핸들러와의 통일을 위해 공백 추가

devholic22 committed:
refactor: Location, Position 표현 수정

- Location 구를 구/군으로 수정
- Position 앞에 좌표 붙여지도록 수정

devholic22 committed:
refactor: ProfileFactory 공백 추가

- 다른 Factory와의 통일을 위해 공백 추가

devholic22 committed:
feat: info 수정 요청 DTO 정의

- info (profile, option, hobbies, styles) 수정 요청 DTO 정의

devholic22 committed:
feat: info 수정 요청 이벤트 정의

- info (profile, option, hobbies, styles) 수정 요청 이벤트 정의
- 관련 이벤트 핸들러 정의
- 기존 이벤트 패키지 분할

devholic22 committed:
feat: 스타일 수정 구현

- 기존 스타일 삭제 메서드 제거, 수정에서만 사용하도록 변경
- 스타일 저장, 수정 메서드만 테스트되도록 수정
- 저장 메서드와 공유하는 코드 메서드화

devholic22 committed:
refactor: StyleServiceTest 이름 변경

- MemberStyleServiceTest로 이름 변경

devholic22 committed:
feat: 취미 수정 구현

- 기존 취미 삭제 메서드 제거, 수정에서만 사용하도록 변경
- 취미 저장, 수정 메서드만 테스트되도록 수정
- 저장 메서드와 공유하는 코드 메서드화
- HobbyService를 MemberHobbyService로 이름 변경

devholic22 committed:
refactor: StyleService 이름 변경

- StyleService를 MemberStyle로 이름 변경

devholic22 committed:
refactor: MemberHobbyServiceTest then 변수 위치 수정

- 리포지터리 조회 변수를 when절로 이동

devholic22 committed:
refactor: MemberStyleServiceTest 오타 수정

- 클래스 이름 Member 중복 삭제

devholic22 committed:
feat: Option 수정 구현

- OptionFactory 수정 Option 생성 메서드 추가, 기존 메서드명 변경
- OptionUpdateRequestFixture 생성
- Option 저장, 수정 메서드만 테스트되도록 수정

devholic22 committed:
feat: InfoService 수정 메서드 작성

- info (profile, option, hobbies, styles) 이벤트 발신 메서드 작성
- 수정 요청 DTO 오류 수정

devholic22 committed:
refactor: Fixture에 SuppressWarnings 추가

- 가독성을 위해 Fixture에 SuppressWarnings 추가

devholic22 committed:
refactor: HobbyInvalid 이름 수정

- 문법에 맞게 클래스 이름 수정

devholic22 committed:
refactor: DrinkInvalid 이름 수정

- 문법에 맞게 클래스 이름 수정

devholic22 committed:
refactor: 클래스 이름 수정

- 문법에 맞지 않는 클래스 이름들 수정

devholic22 committed:
refactor: MemberStyleRepository final 추가

- 메서드 인자 final 추가

devholic22 committed:
refactor: PositionTest 인자 수정

- 예외 검증 시 double 인자로 검증되도록 수정

devholic22 committed:
feat: InfoController 정보 수정 메서드 구현

- InfoController 정보 수정 메서드 구현
- 정보 수정 테스트 구현
- 테스트 문서 반영

devholic22 committed:
refactor: StyleInvalid 이름 수정

- 문법에 맞게 클래스 이름 수정

devholic22 committed:
test: InfoServiceTest 정보 수정 테스트

- InfoServiceTest 정보 수정 테스트 작성
- InfoServiceTest Fixture 생성

devholic22 committed:
docs: InfoController 수정 response field 추가

- InfoController 수정 response field 추가

devholic22 committed:
refactor: InfoControllerWebMvcTest Fixture 추가

- InfoContollerWebMvcTest 조회 Fixture 추가
- 용어 오타 수정

devholic22 committed:
test: InfoServiceTest 조회 테스트 작성

- 테스트 격리 (IntegrationHelper) 추가
- 좌표 요청/응답 Fixture 오류 수정

devholic22 committed:
refactor: ProfileServiceTest ignoringFields 추가

- ProfileServiceTest 저장 검증 시 ignoringFields 추가

devholic22 committed:
refactor: 인터페이스 인자 final 추가

- MemberStyleJpaRepository, OptionJpaRepository final 인자 추가

devholic22 committed:
refactor: OptionFactory 메서드 통일

- OptionFactory 메서드 통일하여 하나의 메서드만으로 생성되도록 수정

devholic22 committed:
feat: ProfileService 수정 메서드 구현

- ProfileService 수정 메서드 구현
- ProfileFactory 메서드 추가
- ProfileServiceTest 저장, 수정만 검증하도록 수정

devholic22 committed:
fix: ProfileEventHandler 서비스 호출 등록

- ProfileEventHandler 수정 이벤트 서비스 호출 등록 수정

devholic22 committed:
refactor: OptionFactory request 추상화

- OptionFactory 메서드 통일을 위해 OptionRequest 인터페이스 등록
- 레코드를 클래스로 수정

devholic22 committed:
refactor: ProfileFactory request 추상화

- ProfileFactory 메서드 통일을 위해 ProfileRequest 인터페이스 및 하위 속성 인터페이스 등록
- 레코드를 클래스로 수정

devholic22 committed:
fix: OptionRequest 구현체 DTO 기본 생성자 추가

- OptionRequest 구현체 DTO 기본 생성자 추가

devholic22 committed:
refactor: Request private 속성 추가

- Request 필드에 private 속성 추가 통일

devholic22 committed:
refactor: Job 조회 메서드명 수정

- 통일성을 위해 조회 메서드명 수정

devholic22 committed:
refactor: HobbyJpaRepository 이름 수정

- 통일성을 위해 MemberHobbyJpaRepository로 이름 수정

devholic22 committed:
refactor: Style 검증 코드 분할

- Style 검증 코드 메서드 분할

devholic22 committed:
refactor: 응답 DTO 공백 추가

- 통일성을 위해 응답 DTO들의 메서드 전 공백 추가

devholic22 committed:
refactor: JpaRepositoryTest ignoringFields 추가

- MemberHobbyJpaRepositoryTest ignoringFields 추가
- MemberStyleJpaRepositoryTest ignoringFields 추가

devholic22 committed:
fix: Style 검증 예외 수정

- Style 검증 예외 수정

eom-tae-in commented on src/main/java/com/atwoz/member/application/info/InfoService.java:
```diff
@@ -0,0 +1,87 @@
+import com.atwoz.member.application.info.dto.StyleUpdateRequest;
+import com.atwoz.member.application.info.dto.StyleWriteRequest;
+import com.atwoz.member.domain.info.InfoRepository;
+import com.atwoz.member.ui.info.dto.InfoSearchResponse;
```
역참조가 발생하는데 구조 변경이 필요해보입니다.

devholic22 replied:
InfoSearchResponse의 역참조를 없애기 위해 Infrastructure로 옮기는 방식으로 해결하였습니다!

eom-tae-in commented on src/main/java/com/atwoz/member/application/info/InfoService.java:
```diff
@@ -0,0 +1,87 @@
+                .stream()
+                .map(StyleWriteRequest::style)
+                .toList();
+    }
```
request.styles().stream(), request.hobbies().stream()이 한 맥락인데 매개변수를 InfoWriteRequest로 받는 이유가 있나요?
한 줄에 체이닝을 한번 사용하기 위해 일부러 분리한 것 처럼 보이네요
request를 넘기지 말고 값을 꺼내서 넘기도록 수정해주세요~

devholic22 replied:
값을 한번 꺼내서 넘기면 불필요한 체이닝을 제거할 수 있겠군요..! 수정 완료하였습니다 :)

eom-tae-in commented on src/main/java/com/atwoz/member/application/info/InfoService.java:
```diff
@@ -0,0 +1,87 @@
+                .stream()
+                .map(StyleUpdateRequest::style)
+                .toList();
+    }
```
이것도 request 값 꺼내서 넘겨주세요~

devholic22 replied:
수정하였습니다 :)

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/option/OptionWriteRequest.java:
```diff
@@ -0,0 +1,28 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+public class OptionWriteRequest implements OptionRequest {
```
OptionUpdateRequest랑 완전 똑같은 클래스네요

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/option/OptionRequest.java:
```diff
@@ -0,0 +1,10 @@
+package com.atwoz.member.application.info.dto.option;
+
+public interface OptionRequest {
```
인터페이스로 나눈 이유가 있나요?
OptionUpdateRequest 같은 곳에서는 해당 메서드를 오버라이드 하지도 않는 것 같아보이고 나눈 이유가 있나 궁금해요

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/profile/location/LocationRequest.java:
```diff
@@ -0,0 +1,7 @@
+package com.atwoz.member.application.info.dto.profile.location;
+
+public interface LocationRequest {
```
이 부분도 그렇고 다른 부분 모두 왜 인터페이스로 분리를 한 건가요??

sosow0212 commented on src/main/java/com/atwoz/member/application/info/hobby/MemberHobbyService.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Transactional
+    public void updateMemberHobbies(final Long memberId, final List<String> hobbyCodes) {
```
변경감지로 업데이트 할 수 있지 않을까요? 쿼리 문제도 그렇고, 도메인 로직을 외부에서 처리하는 것처럼 보이네요!

devholic22 replied:
이것과 관련해서는 고민점이 있습니다.
* MemberHobby, MemberStyle은 memberId와 Hobby / Style이 M:N으로 되어 있는 구조입니다.
* 그렇기 때문에 변경 감지를 하기 위해서는, 기존에 저장된 내역과 새로 들어온 요청들을 하나씩 검사하면서 아래 경우의 수가 존재하게 됩니다.
  * `회원의 취미/스타일 갯수` > `새로 들어온 (수정할) 취미/스타일 갯수`
    * 이 때는 새로 들어온 갯수만큼 변경 감지를 이용하고, (회원의 취미/스타일 갯수) - (새로 들어온 취미/스타일 갯수)만큼 삭제 쿼리가 발생합니다.
  * `회원의 취미/스타일 갯수` = `새로 들어온 (수정할) 취미/스타일 갯수` (이 때는 말씀하신 변경 감지를 이끌어낼 수 있을 듯 합니다.)
  * `회원의 취미/스타일 갯수` < `새로 들어온 (수정할) 취미/스타일 갯수`
    * 이 때는 저장된 갯수만큼 변경 감지를 이용하고, (새로 들어온 취미/스타일 갯수) - (회원의 취미/스타일 갯수)만큼 추가 쿼리가 발생합니다.
* 이렇듯 변경 감지를 위해서는 세 가지에 대해서 각각 구현을 하고, 그렇기에 복잡도가 올라간다는 생각이 들었습니다.
* 그 대신, 기존에 `updateMemberHobbies`를 할 때 작성했던 `deleteHobbiesByMemberId` 메서드가 저장되었던 `취미 갯수만큼 각각 delete` 쿼리가 일어났던 문제를 `@Query("delete from MemberHobby mh where mh.memberId = :memberId")`로 한 번에 삭제하도록 수정했는데, 이런 방식으로 이끌어나가도 될 지 여쭤드립니다..!

sosow0212 commented on src/main/java/com/atwoz/member/application/event/info/StyleWroteEvent.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+@RequiredArgsConstructor
+public class StyleWroteEvent extends Event {
```
파일체인지에 보이지가 않아서 여기에 남깁니다!
application/event 안에 ValidatedLoginEvent는 패키지 참조가 잘못 되어 있는 것 같아요
그리고 이벤트의 패키지는 어디로 가는게 좋을까요?

devholic22 replied:
이벤트 패키지에 대해 다시 고민을 해 보니, 자신이 처리하는 곳에 더 가깝게 위치하는 것이 낫겠다는 생각을 하였습니다.
예시로 `application/profile/dto`로 profileService가 필요로 하는 dto가 위치해 있듯이, `application/profile/event`로 두면 자신과 관련된 이벤트들이 동일한 패키지들이 묶여 더 합리적인 구조를 만들 수 있겠다고 생각하여 수정하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/application/info/option/OptionFactory.java:
```diff
@@ -0,0 +1,22 @@
+
+public class OptionFactory {
+
+    public static Option createOption(final Long memberId, final OptionRequest optionRequest) {
```
도메인에서 처리하면 되는데 Factory 클래스를 만든 이유가 있나요?

devholic22 replied:
인자가 많거나 구현 로직이 복잡한 경우 Factory를 사용하려는 취지로 접근했었습니다.
말씀하신 뒤 Factory에 대해서 생각을 다시 하게 되었고, 그 대신 `정적 팩터리 메서드`로 대체하였습니다.
정적 팩터리도 인자가 너무 많은 것 같아 쓰지 않는 게 낫겠다는 생각을 했었는데, 다시 생각해보니 객체에게 책임을 더 부여해줄 수 있다는 게 느껴져서 수정하였습니다.
가급적 Factory를 사용하지 않는 방향으로 작성하겠습니다..!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/profile/ProfileService.java:
```diff
@@ -0,0 +1,46 @@
+                newProfile.getLocation(),
+                newProfile.getPosition(),
+                newProfile.getJob()
+        );
```
이 부분도 Factory보다 도메인에서 처리하는게 나아보입니다
굳이 객체를 새로 생성하고, 새로 생성한 객체의 값들을 기존 값에 덮어쓰고 있는 것 같은데, 이렇게 되면 update하는 도메인 로직은 결국 외부인 Factory로직에 의존하는 거가 되는 것 같습니다

Factory를 사용하지 말던가, 아니면 사용하고 싶다면 도메인 안에서 사용하는게 낫지 않을까요?

devholic22 replied:
도메인 전용 DTO를 사용하는 방식으로 수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/profile/location/LocationUpdateRequest.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+public class LocationUpdateRequest implements LocationRequest {
+
+        @NotBlank(message = "시/도가 작성되어야 합니다.")
```
여긴 정렬이 이상한 것 같습니다!

devholic22 replied:
레코드로 만들었다가 클래스로 변환하는 과정에서 탭이 유지된 것 같습니다..! 레코드로 복구하였습니다.

sosow0212 commented on src/main/java/com/atwoz/member/application/info/dto/profile/location/LocationWriteRequest.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+public class LocationWriteRequest implements LocationRequest {
+
+        @NotBlank(message = "시/도가 작성되어야 합니다.")
```
여기도 정렬이 이상하네요

sosow0212 replied:
밑에 다른 부분도 이상한 부분이 많은데 확인 부탁드립니다

devholic22 replied:
전부 레코드를 빠르게 클래스로 변환하려다보니 발생한 문제 같습니다.. 바로 수정하였습니다!

sosow0212 commented on src/main/java/com/atwoz/member/application/info/style/MemberStyleService.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Transactional
+    public void updateMemberStyles(final Long memberId, final List<String> styleCodes) {
```
이 부분도 도메인에서 변경감지로 처리되어야하는데 외부에서 처리가 되네요 ㅠㅠ

sosow0212 commented on src/main/java/com/atwoz/member/infrastructure/info/hobby/MemberHobbyRepositoryImpl.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public void saveAll(final List<MemberHobby> memberHobbies) {
+        memberHobbyJpaRepository.saveAll(memberHobbies);
```
다른 부분도 그렇지만, memberHoobyJpaRepository에 해당 메서드들을 만들어줘야하지 않을까요? 기본 메서드가 있더라도 명시해주면 좋은 이유는 무엇일까요?

devholic22 replied:
`인터페이스의 존재 목적`을 더 명확히 하기 위해서라고 생각됩니다.
save, saveAll 등 기본 메서드들은 상위 인터페이스인 JpaRepository, CrudRepository 등에 내장되어 있는데, 이렇게 되면 개발자가 사용하게 되는 인터페이스인 OptionJpaRepository 등에서는 서비스에서 사용하고자 하는 규약 중 일부가 감춰져 있기 때문에 혼동을 줄 수도 있겠다는 생각이 들었습니다.

리포지터리들에 메서드들을 작성하는 방식으로 수정하였습니다. 다만 saveAll의 경우, 재정의한다면 saveAll을 엔티티의 속성으로 인식하는 문제가 발생하여 기존의 부모 인터페이스 메서드를 사용하도록 하였습니다. (saveAll에 대해서는 추가적으로 정확한 원리에 대해 알아봐야 할 것 같습니다.)

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/dto/HobbySearchResponse.java:
```diff
@@ -0,0 +1,15 @@
+import com.atwoz.member.domain.info.hobby.MemberHobby;
+
+public record HobbySearchResponse(
+
```
불필요한 공백이 있습니다

devholic22 replied:
수정 완료하였습니다

sosow0212 commented on src/main/java/com/atwoz/member/ui/info/dto/profile/ProfileSearchResponse.java:
```diff
@@ -0,0 +1,44 @@
+public record ProfileSearchResponse(
+
+        int age,
+
```
굳이 중간 공백은 없어도 될 것 같습니다!

devholic22 replied:
수정 완료하였습니다! Request DTO에서 어노테이션 때문에 공백을 뒀어서 Response DTO에서도 습관적으로 공백을 둔 것 같습니다. 별도의 어노테이션이 없다면 공백을 제거해두도록 하겠습니다.

sosow0212 commented on src/main/java/com/atwoz/member/infrastructure/info/InfoRepositoryImpl.java:
```diff
@@ -0,0 +1,78 @@
+    private Option selectOptionByMemberId(final Long memberId) {
+        return Optional.ofNullable(queryFactory.select(QOption.option)
+                .from(QOption.option)
+                .where(QOption.option.memberId.eq(memberId))
```
.orElseThrow는 서비스에서 처리해줘야할 것 같습니다

devholic22 replied:
수정 완료하였습니다! 조인하는 과정 때문에 리포지터리에서 하는 것으로 생각했었던 것 같습니다.

sosow0212 commented on src/main/java/com/atwoz/member/infrastructure/info/InfoRepositoryImpl.java:
```diff
@@ -0,0 +1,78 @@
+        return new InfoSearchResponse(profileSearchResponse, optionSearchResponse, hobbyCodes, styleCodes);
+    }
+
+    private Profile selectProfileByMemberId(final Long memberId) {
```
QueryDSL 사용하는 부분도 클래스로 분리하는게 좋을 것 같습니다

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/InfoServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+    void 모든_정보를_수정한다() {
+        // given
+        Long memberId = 1L;
+        InfoWriteRequest infoWriteRequest = 정보_생성_요청();
```
무슨 정보인지 명시해주면 좋을 것 같습니다 나중에 많아지면 헷갈릴 것 같아요

devholic22 replied:
수정 완료하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/application/info/MemberHobbyServiceTest.java:
```diff
@@ -0,0 +1,78 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+class MemberHobbyServiceTest {
```
서비스에서 예외 검증이 하나도 안 되어있는데 이런 부분도 검증해주세요

devholic22 replied:
수정 완료하였습니다! 스타일도 동일하게 적용하였습니다.

sosow0212 commented on src/test/java/com/atwoz/member/application/info/OptionServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+        Option exprectedOption = OptionFactory.createOption(memberId, updateRequest);
+
+        OptionWriteRequest writeRequest = 옵션_생성_요청();
+        optionService.writeOption(memberId, writeRequest);
```
서비스에 다른 메서드가 바뀌면 해당 테스트는 변경될 것 같습니다
차라리 하위 계층인 repository를 이용하는 건 어떨까요?

devholic22 replied:
무슨 말씀인지 이해하였습니다! 서비스 테스트인데 다른 메서드를 먼저 호출할 경우 의도치 않게 바뀔 수 있다는 것을 놓쳤었습니다
수정 완료하였습니다

sosow0212 commented on src/test/java/com/atwoz/member/ui/auth/AuthControllerWebMvcTest.java:
```diff
@@ -35,6 +37,9 @@ class AuthControllerWebMvcTest extends MockBeanInjection {
     @Autowired
     private ObjectMapper objectMapper;
 
+    @MockBean
+    private MemberRepository memberRepository;
```
이건 MockBeanInjection에 넣어야할 것 같아요

devholic22 replied:
MemberRepository 자체를 사용하지 않음을 발견하여, 삭제 조치하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/ui/info/InfoControllerWebMvcTest.java:
```diff
@@ -0,0 +1,189 @@
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private InfoService infoService;
```
이것도 MockBeanInjection에 넣어야합니다

devholic22 replied:
수정하였습니다!

sosow0212 commented on src/test/java/com/atwoz/member/ui/info/InfoControllerWebMvcTest.java:
```diff
@@ -0,0 +1,189 @@
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private InfoService infoService;
```
E2E 테스트도 진행해주세요~

devholic22 replied:
작성 완료하였습니다! 처음 작성해 본 E2E 테스트인데, 고칠 점이 있다면 바로 반영하겠습니다 :)

sosow0212 reviewed (changes requested):
수고하셨습니다 몇 부분 확인하고 계층구조랑 코드 스타일 조금만 더 신경써주세요~ 고생하셨습니다잉

devholic22 committed:
refactor: AllArgsConstructor access 단계 수정

- AllArgsConstructor access 단계 PRIVATE으로 모두 수정

devholic22 committed:
refactor: ProfileServiceTest 예외 검증

- ProfileServiceTest 예외 검증 추가
- findMemberById 메서드 위치 수정
- 테스트 메서드명 한글화 통일

devholic22 committed:
refactor: OptionServiceTest 예외 검증

- OptionServiceTest 예외 검증 추가
- findMemberById 메서드 위치 수정
- 테스트 메서드명 한글화 통일

devholic22 committed:
refactor: InfoService MockBeanInjection 이동

- InfoService MockBean을 MockBeanInjection으로 이동

devholic22 committed:
refactor: MemberRepository 의존 삭제

- AuthControllerWebMvcTest MemberRepository 의존 삭제

devholic22 committed:
refactor: DTO 공백 삭제

- DTO들의 공백 삭제

devholic22 committed:
refactor: Request DTO 인터페이스 삭제

- Request DTO 인터페이스 삭제, 레코드로 복구

devholic22 committed:
refactor: Q파일 static import 처리

- Q파일 static import 등록

devholic22 committed:
refactor: 테스트 용어 구분 추가

- 테스트 문맥을 파악할 수 있도록 회원 용어 추가되도록 수정

devholic22 committed:
test: InfoController 인수 테스트 작성

sosow0212 committed:
refactor: 테스트 메서드 수정

devholic22 committed:
refactor: Profile, Option, MemberHobby, MemberStyle Fixture 작성

- 코드 중복 제거를 위해 Profile, Option, MemberHobby, MemberStyle Fixture 추가

devholic22 committed:
refactor: 인수 테스트 given Fixture 이용

- 인수 테스트 given 절의 데이터를 Fixture로 변경

devholic22 committed:
refactor: 테스트 요청 Fixture 구조 수정

- API 문서에서 사용되는 데이터와 동일하게 이용되도록 수정
- RequestFixture를 만든 뒤 생성/수정 데이터가 사용되도록 수정

devholic22 committed:
refactor: Querydsl 구현체 분리

- InfoRepositoryImpl과 JPAQueryFactory 분리되도록 수정

devholic22 committed:
refactor: MemberHobby, MemberStyle 삭제 쿼리 개선

- MemberHobby, MemberStyle 삭제 쿼리 개선

devholic22 committed:
test: InfoController 인수 테스트 추가 작성

- 회원 정보 조회 예외 검증 추가
- 회원 정보 수정 검증 추가
- 회원 정보 수정 예외 검증 추가

devholic22 committed:
refactor: InfoSearchResponse 예외 이동

- InfoSearchResponse 예외를 서비스 계층으로 이동
- 관련 예외 추가

devholic22 committed:
refactor: InfoServiceTest 이벤트 mock 처리

- InfoServiceTest 이벤트 mock 처리 방식으로 수정
- ApplicationEventPublisher mock 등록
- 조회 전용 fake 리포지터리 등록 (값 저장 메서드가 없기에 고정적으로 반환하도록 설정)

devholic22 committed:
refactor: 수정 테스트 시 리포지터리를 이용하도록 수정

- 수정 테스트의 기본값을 만들 때 리포지터리를 이용하도록 수정
- OptionServiceTest 변수명 오타 수정
- ProfileFixture 수정 전용 객체 메서드 정의
- OptionFixture 수정 전용 객체 메서드 정의

devholic22 committed:
refactor: OptionFactory 제거

- OptionFactory 제거
- Option 정적 팩터리 메서드로 대체

devholic22 committed:
refactor: Profile, Option 업데이트 방식 변경

- 새로운 엔티티를 받아서 자신의 도메인 로직에서 수정하도록 변경

devholic22 committed:
refactor: Option 내부 DTO 생성

- Option에서 생성/수정을 할 때 도메인 전용 DTO를 이용하도록 수정
- 요구하는 곳이 도메인이므로 도메인 내부에 DTO 패키지 정의
- 테스트 코드 수정

devholic22 committed:
refactor: ProfileFactory 삭제

- Profile에서 생성/수정을 할 때 도메인 전용 DTO를 이용하도록 수정
- 요구하는 곳이 도메인이므로 도메인 내부에 DTO 패키지 정의
- 테스트 코드 수정
- Fixture 구조 수정
- DTO 패키지 분리

devholic22 committed:
refactor: BodySearchResponseFixture 클래스로 복구

- BodySearchResponseFixture 클래스화

devholic22 committed:
refactor: Profile, Option 조인 조회 방식으로 수정

- Profile, Option 조회 시 한 번에 조회되도록 조인
- ProfileAndOptionSearchResponse 정의

devholic22 committed:
refactor: 이벤트 패키지 수정

- 이벤트 패키지 구조 수정 (도메인 중심으로 수정)

devholic22 committed:
refactor: DTO 패키지 수정

- DTO들이 각 도메인, 애플리케이션 하위 패키지로 들어가도록 수정

devholic22 committed:
refactor: 이벤트 패키지 위치 조정

- 각 서비스 단위로 이벤트가 관리되도록 패키지 위치 조정

devholic22 committed:
refactor: 응답 DTO 위치 조정

- 응답 DTO 위치를 인프라로 수정 (만드는 곳에서 관리되도록 수정)

devholic22 committed:
refactor: Fixture 구조 수정

- Fixture 내부에서도 계층을 나누도록 수정
- 프로덕션 구조와 테스트 구조가 동일하게 관리되도록 수정

devholic22 committed:
refactor: InnerProfile DTO 위치 수정

- InnerProfile DTO를 도메인 단에 두도록 수정

devholic22 committed:
refactor: 서비스 테스트 패키지 수정

- 서비스 테스트 패키지 구조 수정

devholic22 committed:
test: 취미 서비스 단계의 예외 검증 추가

- MemberHobbyService 예외 검증 추가

devholic22 committed:
test: 스타일 서비스 단계의 예외 검증 추가

- MemberStyleService 예외 검증 추가

devholic22 committed:
refactor: 도메인 테스트 패키지 수정

- 도메인 패키지 구조 수정

devholic22 committed:
refactor: JPA 리포지터리 메서드들이 명시되도록 수정

- JPA 리포지터리 메서드들이 명시되도록 수정

devholic22 committed:
fix: findAll 메서드 재정의 삭제

- findAll 메서드는 기존 부모 메서드를 사용하도록 복구

devholic22 committed:
refactor: Fixture static import 추가

- Fixture 사용 시 이름이 충돌되지 않을 경우 static import 되도록 수정

devholic22 committed:
test: InfoQueryRepository 테스트 추가

- InfoQueryRepository 테스트 추가
- 예외를 방지하기 위해 JPAQueryFactory 수동 빈 등록 환경 추가

devholic22 committed:
refactor: Job 예외 패키지 수정

- Job 예외 패키지 구조 수정 (job 패키지 안에 들어가도록 수정)

devholic22 committed:
refactor: Option, Body 정적 팩터리 메서드 수정

- 정적 팩터리 메서드 사용 시 변수를 밖으로 이끌어내지 않도록 수정

devholic22 committed:
refactor: InnerOptionUpdateRequest 공백 추가

- InnerOptionUpdateRequest 정적 팩터리 메서드 공백 추가

devholic22 committed:
refactor: Profile updateContents 삭제

- updateContents를 다른 곳에서 사용하지 않으므로 삭제
- 관련 테스트 수정

devholic22 committed:
refactor: InnerProfileUpdateRequest Fixture 추가

- InnerProfileUpdateRequest Fixture 등록

eom-tae-in reviewed (approved):
고생하셨습니다

devholic22 committed:
refactor: InfoService 메서드 체이닝 최소화

- InfoService 메서드 체이닝이 최소화되도록 수정

sosow0212 reviewed (approved):
태인이가 말한 부분 고쳐주시면 될 것 같습니다.
작업하시느라 수고하셨습니다~
